crates/saya/core/src/prover/program_input.rs

```diff
@@ -0,0 +1,220 @@
+use anyhow::bail;
+use katana_primitives::contract::ContractAddress;
+use katana_primitives::state::StateUpdates;
+use katana_primitives::trace::{EntryPointType, TxExecInfo};
+use katana_primitives::transaction::L1HandlerTx;
+use katana_primitives::utils::transaction::compute_l1_message_hash;
+use starknet::core::types::FieldElement;
+
+use super::state_diff::state_updates_to_json_like;
+
+/// Based on https://github.com/cartridge-gg/piltover/blob/2be9d46f00c9c71e2217ab74341f77b09f034c81/src/snos_output.cairo#L19-L20
+/// With the new state root computed by ten prover.
+pub struct ProgramInput {
+    pub prev_state_root: FieldElement,
+    pub block_number: FieldElement,
+    pub block_hash: FieldElement,
+    pub config_hash: FieldElement,
+    pub message_to_starknet_segment: Vec<MessageToStarknet>,
+    pub message_to_appchain_segment: Vec<MessageToAppchain>,
+    pub state_updates: StateUpdates,
+}
+
+pub fn extract_messages(
+    exec_infos: &Vec<TxExecInfo>,
+    mut transactions: Vec<&L1HandlerTx>,
+) -> (Vec<MessageToStarknet>, Vec<MessageToAppchain>) {
+    let message_to_starknet_segment = exec_infos
```

Comment:
I'm not sure if this one is fully recursive. Because `inner_calls` may also have other `inner_calls` at their turn etc...

Could we try an approach like it was done here:
https://github.com/dojoengine/dojo/blob/64fa43c6551bc5fe4eff2676723f066ccb4e12cd/crates/katana/executor/src/implementation/sir/output.rs#L100-L130

Reply:
I think my stack approach would work with recursive calls, the proposed function is cleaner, so changed the code to use it.